### ISSUE TITLE
Multi-repo support 🌲

### ISSUE THREAD

**Issue by edvald** [garden-io/garden#162]
Supporting Garden projects that span multiple repositories will be a critical feature for many users.

To support that properly, we need to be able to do the following:
1. Reference repositories via git URLs, in the project configuration and/or as sources in module configs.
2. Automatically clone those repos and keep them updated.
3. Locally link a repo to a local clone, so the user can work on multiple repos without circling through the remote repo.

Each of these has some inherent complexities and needs careful design and implementation.

**1 - Referencing external sources**

We can imagine a few different ways to link external sources (i.e. repos) to a project.
One is to allow a list of sources to be specified in the project config, which will all be pulled in when scanning modules, and then treated as part of the project tree. It could look something like this:

```yaml
project:
  name: my-project
  # ...
  sources:
  - name: other-repo
    location: git://github.com/my/repo#master
  # ...
```

We'd then pull each of the referenced repos (unfortunately having to poll for updates changes every time) and then scan for `garden.yml` files in each of the clones, much like we do currently in the project directory.

Another way could be to add something like an `external` or `sources` _top-level_ key to the `garden.yml` file, which would have a similar format as the above `sources` key. A benefit of that would be that you wouldn't have to have all the external references in the top-level project config (and they could even cascade across multiple repos), but otherwise it would work functionally the same.

And a third option would be to have an optional `source` key in module configs, so you'd essentially define the `garden.yml` for a module in the main project repo, and then just refer to a repo for its source code.

The above aren't actually mutually exclusive, but I'd prefer to pick a single canonical way to do this, at least to start. I'd personally lean towards having a top-level `sources` key in the `garden.yml` file.

**2 - Cloning and syncing linked repos**

Cloning the referenced repos should be fairly straightforward. I guess the only question might be _where_ we clone them to. I think something like `.garden/sources/<source name>` would make sense.

Keeping the clone in sync is slightly more tricky, particularly when auto-reloading, but also not super complicated. We'd probably need to settle for basic polling for changes, possibly with a configurable poll rate. You might also want to disable automatic updates, if we provide a command to manually fetch latest changes from remote (e.g. `garden update-sources`)

**3 - Link local working copies**

Of course you'll want to be able to work on linked repos without having to push changes to the remote all the time. I figure we could add a command to do that locally (since you don't want to hard-code a reference to a local directory in the `garden.yml` config). Something like:

    garden link-source name-of-source ../path/to/local/clone

This "link" would then be stored in your local config and when present the framework will simply look to the local path for the linked sources. This would mean that the user is then responsible for having their local working copy up to date, on the correct branch etc. (because we can't reliably take care of that, in case the working copy is modified/dirty etc.). And we'd of course need a corresponding `unlink` command.

I imagine it could get a little thorny for users to separately commit changes in multiple repos etc. but then again they likely already have that problem...

I think that mostly covers what needs to happen. Am I missing something?

**Comment by eysi09** [comment]
This sounds pretty reasonable and I like the workflow on user's end.

I'm a little confused though as to the difference between option one and option two in section:  **1 - Referencing external sources**.

**Comment by edvald** [comment]
Just as I hit the button, another way occurs to me. Thinking aloud here in terms of the details, but, we might want to support nested projects. This would actually be rather neat if you have multiple teams, but might add a fair amount of complexity to the framework (e.g. how do environment configurations cascade?).

But, we could sidestep that complexity for a bit, just add the top-level `project.sources` key instead of allowing it in any config file, but allow for the flexibility of nested projects sometime later. I figure that's also the simplest first thing to implement, which is always a plus.

**Comment by eysi09** [comment]
Yes, I can see the value in being able to declare dependencies at the module level. Of course nested projects would enable that, but it feels overly complex (and semantically wrong) if you just need your module to import another module.

On the other hand, one could argue that a self contained module that depends on other modules is really just a project. If you have a team working on just that module (and perhaps it's dependencies) in isolation it would need a project config to run it anyway. 

So, B + nested projects might be the way to go. B would make things more explicit and benefit smaller teams and nested projects provide the flexibility larger teams need.

In that respect, perhaps nested module dependencies where the parent module is itself not a project is kind of an edge case that we shouldn't be supporting right away.

**Comment by eysi09** [comment]
Resolved in #219
